src/vmm/src/devices/virtio/iovec.rs

```diff
@@ -612,3 +612,117 @@ mod tests {
         assert_eq!(sub.iovecs[1].iov_base, iovec.vecs[1].iov_base);
     }
 }
+
+#[cfg(kani)]
+mod verification {
+    use std::mem::ManuallyDrop;
+
+    use libc::{c_void, iovec};
+
+    use super::{IoVecBuffer, IoVecBufferMut};
+
+    // Maximum memory size to use for our buffers. For the time being 1KB.
+    const GUEST_MEMORY_SIZE: usize = 1 << 10;
+
+    // Maximum number of descriptors in a chain to use in our proofs. The value is selected upon
+    // experimenting with the execution time. Typically, in our virtio devices we use queues of up
+    // to 256 entries which is the theoretical maximum length of a `DescriptorChain`, but in reality
+    // our code does not make any assumption about the length of the chain, apart from it being
+    // >= 1.
+    const MAX_DESC_LENGTH: usize = 5;
+
+    fn create_iovecs(mem: *mut u8, size: usize) -> (Vec<iovec>, usize) {
+        let nr_descs: usize = kani::any_where(|&n| n <= MAX_DESC_LENGTH);
+        let mut vecs: Vec<iovec> = Vec::with_capacity(nr_descs);
+        let mut len = 0usize;
+        for _ in 0..nr_descs {
+            // The `IoVecBuffer(Mut)` constructors ensure that the memory region described by every
+            // `Descriptor` in the chain is a valid, i.e. it is memory with then guest's memory
+            // mmap. The assumption, here, that the last address is within the memory object's
+            // bound substitutes these checks that `IoVecBuffer(Mut)::new() performs.`
+            let addr: usize = kani::any();
+            let iov_len: usize =
+                kani::any_where(|&len| matches!(addr.checked_add(len), Some(x) if x <= size));
+            let iov_base = unsafe { mem.offset(addr.try_into().unwrap()) } as *mut c_void;
+
+            vecs.push(iovec { iov_base, iov_len });
+            len += iov_len;
+        }
+
+        (vecs, len)
+    }
+
+    impl kani::Arbitrary for IoVecBuffer {
+        fn any() -> Self {
+            // We only read from `IoVecBuffer`, so create here a guest memory object, with arbitrary
+            // contents and size up to GUEST_MEMORY_SIZE.
+            let mut mem = ManuallyDrop::new(kani::vec::exact_vec::<u8, GUEST_MEMORY_SIZE>());
+            let (vecs, len) = create_iovecs(mem.as_mut_ptr(), mem.len());
+            Self { vecs, len }
+        }
+    }
+
+    impl kani::Arbitrary for IoVecBufferMut {
+        fn any() -> Self {
+            // We only write into `IoVecBufferMut` objects, so we can simply create a guest memory
+            // object initialized to zeroes, trying to be nice to Kani.
+            let mem = unsafe {
+                std::alloc::alloc_zeroed(std::alloc::Layout::from_size_align_unchecked(
```

Comment:
could you please help me understand why the unchecked was chosen over from_size_align?

Reply:
Because all the correctness conditions from here: https://doc.rust-lang.org/core/alloc/struct.Layout.html#method.from_size_align are upheld:

1. 16 is not be zero,
2. 16 is a power of two,
3. `GUEST_MEMORY_SIZE` is aligned to 16 and doesn't overflow `isize`.

I could use `from_size_align` and then `.unrwap` which is exactly same thing.